Dev

evgeniiburdin opened the pull request.

Commit:
task1.2.1.1

Commit:
task1.2.1.2

Commit:
task.1.2.2.1

Commit:
task1.2.2.2

Commit:
task1.2.3.1

Commit:
task1.2.3.2

Commit:
task1.2.5.1

Commit:
task1.3.2.1

Commit:
task1.3.2.2

Commit:
task1.3.3.1

Commit:
task1.3.3.2

Commit:
task1.3.3.3

Commit:
task1.3.4.1

Commit:
task1.3.4.2

Commit:
task1.3.4.3

Commit:
task1.4.1.1

Commit:
task1.4.1.2

Commit:
changed folder structure for 4.variables

Commit:
task1.4.3.1

Commit:
task1.4.5.1

Commit:
task1.5.1.1

Commit:
task1.5.1.2

Commit:
task1.5.2.1

Commit:
task1.5.2.2

Commit:
6.constants

Commit:
7.primitives

inadenenko commented on course1/3.base_program/4.external_import/task1.3.4.1/main.go:
```diff
@@ -0,0 +1,194 @@
+
+import (
+	"fmt"
+	"github.com/shopspring/decimal"
```
imports should be sorted into groups:

```
import (
    "database/sql"
    "io"
    "strconv"

    "golang.org/x/net/context"

    "yourDirectory/foo/bar"
    "yourDirectory/foo/baz"
)
```

inadenenko commented on course1/3.base_program/4.external_import/task1.3.4.1/main.go:
```diff
@@ -0,0 +1,194 @@
+}
+
+func DecimalSum(a, b string) (string, error) {
+	a_dec, err := decimal.NewFromString(a)
```
no snake_case, use camelCase only in vars

inadenenko commented on course1/3.base_program/4.external_import/task1.3.4.1/main.go:
```diff
@@ -0,0 +1,194 @@
+	}
+	if a_dec.GreaterThan(b_dec) {
+		return true, nil
+	} else {
```
else is not needed here

inadenenko commented on course1/3.base_program/4.external_import/task1.3.4.1/main.go:
```diff
@@ -0,0 +1,194 @@
+	if a_dec.LessThan(b_dec) {
+		return true, nil
+	} else {
+		return false, nil
```
else is redundant

inadenenko commented on course1/3.base_program/4.external_import/task1.3.4.1/main.go:
```diff
@@ -0,0 +1,194 @@
+	if a_dec.Equal(b_dec) {
+		return true, nil
+	} else {
+		return false, nil
```
good practice:
check negative result first, then continue your flow for positive income

```
if !aDec.Equal(bDec) {
		return false, nil
	} 

return true, nil
```

inadenenko commented on course1/3.base_program/4.external_import/task1.3.4.3/main.go:
```diff
@@ -0,0 +1,23 @@
+
+import (
+	"fmt"
+	"github.com/icrowley/fake"
```
sort imports

inadenenko commented on course1/3.base_program/4.external_import/task1.3.4.3/main.go:
```diff
@@ -0,0 +1,23 @@
+		fake.StreetAddress(),
+		fake.Phone(),
+		fake.EmailAddress())
+
```
remove empty line

inadenenko commented on course1/4.variables/3.var_global/task1.4.3.1/main.go:
```diff
@@ -0,0 +1,40 @@
+)
+
+func main() {
+
```
remove new line

inadenenko commented on course1/4.variables/3.var_global/task1.4.3.1/main.go:
```diff
@@ -0,0 +1,40 @@
+	fmt.Println(year)
+	fmt.Println(month)
+	fmt.Println(day)
+
```
remove new line

inadenenko commented on course1/4.variables/5.var_pointer/task1.4.5.1/main.go:
```diff
@@ -0,0 +1,37 @@
+import "fmt"
+
+func main() {
+
```
remove new line

inadenenko commented on course1/4.variables/5.var_pointer/task1.4.5.1/main.go:
```diff
@@ -0,0 +1,37 @@
+	changeBool(&b)
+
+	fmt.Println(i, " ", f, " ", s, " ", b)
+
```
remove new line

inadenenko commented on course1/5.pointers/1.pointer_reference/task1.5.1.1/main.go:
```diff
@@ -0,0 +1,43 @@
+}
+
+func IsPrime(number int) *bool {
+	var isPrime bool = true
```
variable is not needed

```
func isPrime(number int) bool {
    if number <= 1 {
        return false
    }
    for i := 2; i < number; i++ {
        if number % i == 0 {
            return false
        }
    }
    return true
}
```

inadenenko commented on course1/6.constants/2.constant_group/task1.6.2.1/main.go:
```diff
@@ -0,0 +1,24 @@
+import "fmt"
+
+const (
+	_    = iota * 2
```
should be done with one iota

```
const (
	id1 = 2 * (iota + 1)
	id2
	id3
	id4
	id5
	id6
	id7
	id8
	id9
	id10
	id11
	id12
	id13
)
```

Commit:
Requested changes completed

task1.3.4.1 -> 	1. imports got sorted into groups
		2. changed snake_case var names to camelCase versions
		3. removed redundant 'else' statements
		4. changed order of checking for negative/positive results in 'if' statements
			in the context of function flow

task1.4.3.1 ->	1. removed redundant newlines

task1.4.5.1 ->	1. removed redundant newlines

task1.5.1.1 ->	1. removed redundant 'isPrime' var in 'IsPrime()'

task1.6.2.1 ->	1. removed redundant iota calls, replaced with single one

Commit:
РазChapter 8: Functions

Commit:
9.test_uint

Commit:
10.benchmark

inadenenko commented on course1/7.primitives/3.int_bitwise/task1.7.3.2/main.go:
```diff
@@ -0,0 +1,25 @@
+
+import "fmt"
+
+var Permissions = map[int]string{
```
no need to do make this map variable global and public

inadenenko commented on course1/7.primitives/4.types_float/task1.7.4.2/main.go:
```diff
@@ -0,0 +1,20 @@
+	client := dinero.NewClient("c91c02817bc54e62b1fa61117e0e3c12", from, 20*time.Minute)
+	rsp, err := client.Rates.Get(to)
+	if err != nil {
+		fmt.Println(err)
```
get used to wrapping errors, especially when they come from a third party dependency
```
if err != nil {
		fmt.Errorf("get exchange rates: %w", err)
```

inadenenko commented on course1/8.func/4.func_return/task1.8.4.1/main.go:
```diff
@@ -0,0 +1,17 @@
+}
+
+func DivideAndRemainder(a, b int) (int, int) {
+	defer func() {
```
bad practice to use recover
just check b for zero and exit in such case

inadenenko commented on course1/8.func/5.func_err/task1.8.5.1/main.go:
```diff
@@ -0,0 +1,30 @@
+	initValue, finalValue := "10.0", "15.0"
+	percentChange, err := CalculatePercentageChange(initValue, finalValue)
+	if err != nil {
+		panic(err)
```
throw panic only in exceptional cases
just print the error and return here

inadenenko commented on course1/8.func/5.func_err/task1.8.5.2/main.go:
```diff
@@ -0,0 +1,28 @@
+
+func main() {
+	disc1, err := CheckDiscount(1488, 0.64)
+	if err != nil {
```
do we need `else` here?
checking error for nil is enough

inadenenko commented on course1/8.func/5.func_err/task1.8.5.2/main.go:
```diff
@@ -0,0 +1,28 @@
+	disc2, err := CheckDiscount(1990, 0.48)
+	if err != nil {
+		fmt.Println(err)
+	} else {
```
remove else

Commit:
11.type_composite(modules 1-4)

Commit:
11.type_composite(modules 5-7)

Commit:
11.type_composite(modules 8-10)

Commit:
12.package

inadenenko commented on course1/11.type_composite/10.composite_interface/task1.11.10.1/main.go:
```diff
@@ -0,0 +1,31 @@
+	case nil:
+		return "Пустой интерфейс"
+	case []int:
+
```
remove empty strings

inadenenko commented on course1/11.type_composite/5.composite_struct/task1.11.5.1/main.go:
```diff
@@ -0,0 +1,38 @@
+
+import (
+	"fmt"
+	"github.com/icrowley/fake"
```
sort imports

inadenenko commented on course1/11.type_composite/5.composite_struct/task1.11.5.2/main.go:
```diff
@@ -0,0 +1,33 @@
+
+import (
+	"fmt"
+	"github.com/icrowley/fake"
```
sort imports

inadenenko commented on course1/11.type_composite/6.struct_composition/task1.11.6.1/main.go:
```diff
@@ -0,0 +1,56 @@
+func (order *Order) AddDish(dish Dish) {
+	order.Dishes = append(order.Dishes, dish)
+}
+func (order *Order) RemoveDish(dish Dish) {
```
add new line between functions

inadenenko commented on course1/11.type_composite/7.composite_map/task1.11.7.2/main.go:
```diff
@@ -0,0 +1,23 @@
+
+import "fmt"
+
+func mergeMaps(map1, map2 map[string]int) map[string]int {
```
what if two maps have same keys?
you should implement logic to count overall value across two maps for same key

inadenenko commented on course1/11.type_composite/7.composite_map/task1.11.7.3/main.go:
```diff
@@ -0,0 +1,16 @@
+
+func createUniqueText(text string) string {
+	words := strings.Fields(text)
+	occurrences := make(map[string]bool)
```
you can do it without bool value, just using empty struct as value
map[string]struct{}

inadenenko commented on course1/11.type_composite/7.composite_map/task1.11.7.5/main.go:
```diff
@@ -0,0 +1,15 @@
+			resultWords = append(resultWords, word)
+		}
+	}
+	return strings.Join(resultWords, " ")
```
add new line after loop

inadenenko commented on course1/11.type_composite/7.composite_map/task1.11.7.5/main.go:
```diff
@@ -0,0 +1,15 @@
+
+import "strings"
+
+func filterSentence(sentence string, filter map[string]bool) string {
```
also no need to use bool here
just use empty struct to determine whether key exists in map

inadenenko commented on course1/11.type_composite/7.composite_map/task1.11.7.7/main.go:
```diff
@@ -0,0 +1,19 @@
+
+func getUniqueUsers(users []User) []User {
+	uniqueUsers := make([]User, 0, len(users))
+	nicknamesUnique := make(map[string]bool)
```
no need to use bool as value
use struct{}

Commit:
Required changes implemented

inadenenko commented on course1/11.type_composite/8.composite_string/task1.11.8.4/main.go:
```diff
@@ -0,0 +1,9 @@
+package main
+
+func concatStrings(xs ...string) string {
+	str := ""
```
you better use strings.Builder type for string concatenation
`+` creates new string with every operation

Commit:
Requsted changes implemented
